src/merkle/merkle_mountain_range/full.rs

```diff
@@ -0,0 +1,357 @@
+        }
+    }
+}
+
+impl DoubleEndedIterator for BitPositionIterator {
+    fn next_back(&mut self) -> Option<<Self as Iterator>::Item> {
+        // trailing_zeros is compute with the intrinsic ctlz. [Rust 1.67.0] x86 uses the `bsr`
+        // instruction. AArch64 uses the `clz` instruction.
+        let zeros = self.value.leading_zeros();
+
+        if zeros == usize::BITS {
+            None
+        } else {
+            let bit_position = usize::BITS - zeros - 1;
+            let mask = 1 << bit_position;
+            self.value ^= mask;
+            Some(bit_position)
+        }
+    }
+}
+
+pub struct Accumulator {
+    /// The number of leaves is used to differentiate accumulators that have the same number of
+    /// peaks. This happens because the number of peaks goes up-and-down as the structure is used
+    /// causing existing trees to be merged and new ones to be created. As an example, every time
+    /// the MMR has a power-of-two number of leaves there is a single peak.
+    ///
+    /// Every tree in the MMR forest has a distinct power-of-two size, this means only the right
+    /// most tree can have an odd number of elements (1). Additionally this means that the bits in
+    /// `leaves` conveniently encode the size of each individual tree.
+    ///
+    /// Examples:
+    ///
+    ///    Example 1: With 5 leaves, the binary 0b101. The number of set bits is equal the number
+    ///    of peaks, in this case there are 2 peaks. The 0-indexed least-significant position of
+    ///    the bit determines the number of elements of a tree, so the rightmost tree has 2**0
+    ///    elements and the left most has 2**2.
+    ///
+    ///    Example 2: With 12 leaves, the binary is 0b1100, this case also has 2 peaks, the
+    ///    leftmost tree has 2**3=8 elements, and the right most has 2**2=4 elements.
+    pub leaves: usize,
+
+    /// All the peaks of every tree in the MMR forest. The peaks are always ordered by number of
+    /// leaves, starting from the peak with most children, to the one with least.
+    pub forest_peaks: Vec<Word>,
+}
+
+/// A fully materialized Merkle Mountain Range, with every tree in the forest and all their
+/// elements.
+///
+/// Since this is a full representation of the MMR, elements are never removed and the MMR will
+/// grow roughly `O(n**2)` in number of leaf elements.
+pub struct FullMerkleMountainRange {
+    /// Refer to the `Accumulator` documentation for details of the semantics of this value.
+    pub leaves: usize,
+
+    /// Contains every element of the forest.
+    ///
+    /// The trees are in postorder sequential representation. This representation allows for all
+    /// the elements of every tree in the forest to be stored in the same sequential buffer. It
+    /// also means new elements can be added to the forest, and merging of trees is very cheap with
+    /// no need to copy elements.
+    nodes: Vec<Word>,
+}
+
+impl FullMerkleMountainRange {
```

Comment:
It might be worth adding a `root()` method to `FullMerkleMountainRange` that computes the root.  Then adding some test cases to ensure expected behaviour.

Reply:
I referring to the concept of "bagging the peaks".  Where you take the roots of the sub trees and combine them in their own merkle tree such that you have a single commitment for the whole MMR.  See https://github.com/opentimestamps/opentimestamps-server/blob/master/doc/merkle-mountain-range.md.

Reply:
Ah, I see! I thought about it in the beginning and decide not to go in that direction because it breaks the low update frequency property of the accumulator. It should also be super easy to achieve that by either creating a new merkle tree or by hashing the sequence of peaks on the accumulator implemented here.

With that said, I don't have any strong opinions here. I can add something like this if you want, but I need some of the properties/use-cases you're looking for

Reply:
> The use case would be to include a succinct MMR commitment (1 Word) in the block header for the notes DB.

For this use case I think the best approach would be to construct a merkle tree from the accumulator. how does that sound to you?

>  Having said that it would be interesting to know how the number of peaks scales with the number of leaves in the worst case.

The worst case scenario for the number of peaks is when the number of _leaves_ (not nodes) in the structure is $2^{k}-1$, it is one element short of becoming a single tree, and it has exactly `k` peaks

Reply:
> For this use case I think the best approach would be to construct a merkle tree from the accumulator. how does that sound to you?

I'd go with sequential hash or something similar as this will minimize complexity inside Miden assembly. The way it will work there is that this short (single word) commitment will be provided via public inputs, and the full set of peaks will be provided non-deterministically via the advice provider.

Reply:
> I'd go with sequential hash or something similar as this will minimize complexity inside Miden assembly

If we create a merkle tree of the peaks, doesn't that mean that from the miden assembly we just need to provide a single merkle path that goes from the MMR to the Merkle root? wouldn't that be less code and maybe faster?

Reply:
Possibly - I think it depends on the specific context. For example, might be more efficient when we have only a single note, but if a transaction involves a dozen of notes inclusion proofs for each involved peak might add up.

In case of a sequential hash, we can use `adv_pipe` instruction to "unhash" the peaks pretty efficiently and we'd need to do this only once for each transaction.

btw, this is what I assumed would be done with `pack` and `unpack` procedures I described [here](https://github.com/0xPolygonMiden/crypto/issues/51#issuecomment-1426828580).

Reply:
got it, thanks!